packages/celer-web-app/src/ui/components/Settings/SettingsConfig.tsx

```diff
@@ -0,0 +1,184 @@
+import produce from "immer";
+import React, { FunctionComponentElement } from "react";
+import { SplitType } from "core/compiler";
+import { Setting } from "core/context";
+import { MapDisplayModes, MapValues, Themes, ThemeValues } from "core/settings";
+import { SettingLabel, SettingToggle, SettingProps, SettingDropdown } from "./SettingsMenuItems";
+
+export const StringToSetting: {[key: string]: React.FC<SettingProps>} = {
+	settinglabel: SettingLabel,
+	settingtoggle: SettingToggle,
+	settingdropdown: SettingDropdown
+};
+
+export type ConfigSetting = {
+	component: string,
+	text: string,
+	action: (draft: Setting) => void,
+	value?: (setting: Setting) => boolean,
+	values?: string[],
+	children?: ConfigSetting[],
+	actionWithValue?: (setting: number) => (draft: Setting) => void,
+	getIndex?: (setting: Setting) => string,
+}
+
+export function render(config: ConfigSetting, setting: Setting, setSetting: (setting: Setting) => void): FunctionComponentElement<SettingProps>
+{
+	if(config.actionWithValue !== undefined && config.getIndex !== undefined)
+	{
+		return React.createElement(
+			StringToSetting[config.component],
+			{
+				text: config.text,
+				action: () => setSetting(produce(setting, config.action)),
+				value: config.value?.(setting),
+				values: config.values,
+				selectedIndex: config.values?.indexOf(config.getIndex(setting)),
+				actionWithValue: config.actionWithValue,
+				actionWithValueUpdate: (func: (draft: Setting) => void) => setSetting(produce(setting, func)),
```

Comment:
I recently learned you can simply do `setSetting(produce(func))`, which is the preferred way since it will always update based on the latest state